### ISSUE TITLE
How to manipulate the z vector in pix2pix

### ISSUE THREAD

**Issue by kaihuchen** [phillipi/pix2pix#35]
@phillipi  
I am trying to use the z vector created in pix2pix  as a latent representation for my research on  various applications related to cGAN.  I use mainly test.lua as my main reference. I need to:

1. Retrieve the z vector out of the model, which I believe should be netG:get(21).output. Is this correct?
2. Bypass the encoder part of the generator to inject a z vector of my choice, and have the decoder generates an image out of this z vector of mine.  

I am not yet sufficiently proficient with Lua/Torch to be certain of what I am doing, and would very much appreciation any pointer for achieving the above.

**Comment by phillipi** [comment]
There are a few options for the z vector. `netG:get(21).ouptut` will get the activations of the innermost layer but before batchnorm. However, you may want to experiment with the codes on a few surrounding layers as well. With the u-net architecture, it's not clear if the "bottleneck" layer (21) will have nice properties, since not all information has to pass through this layer -- in principle, layer 21 can be entirely ignored.

With nngraph, I'm not sure the simplest way to inject noise and just run the generator. One option is to define a new network, equal to the u-net decoder, and copy over the weights from the u-net, like suggested here: https://groups.google.com/forum/#!topic/torch7/KTKQK6uorQ8

Note that you will have to inject noise for each skip connection if you use a u-net. I'd also try removing the skip connections and making the net shallower. With a smaller encoder-decoder like this, I've had some success generating from random zs.

**Comment by kaihuchen** [comment]
@phillipi 

I have investigated this further and found that as you have mentioned, the bottleneck layer does get ignored to a large extent in the U-Net architecture. This is verified by mangling the bottleneck layer and still observing that the output image changes little.

This means that for my purposes I will have to use the encoder-decoder architecture in the generator for both training and testing, otherwise I won't be able to use the bottleneck vector as a meaningful representation for the images. However, I am having problem with running pix2pix in the encoder-decoder architecture, where for the output images I am always seeing a solid-red square in the monitoring UI (through *th -ldisplay.start 8000 0.0.0.0*), and a solid-gray square in the test result (created through test.lua). I wonder if the encoder-decoder architecture is currently functional.
